GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs

```diff
@@ -0,0 +1,176 @@
+            outputStream.Flush();
+            outputStream.Close();
+            context.Response.Close();
+            Stop();
+        }
+
+        /// <summary>
+        /// Returns a key value pair corresponding to the authorization response sent from OAuth2 authorization page.
+        /// Logs error message and throws ArgumentException if the uri contains invalid params.
+        /// </summary>
+        /// <param name="uri">The uri of the incoming request</param>
+        /// <exception cref="ArgumentException">Exception thrown if the uri contains invalid params.</exception>
+        internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
+        {
+            if (!UriContainsValidQueryParams(uri))
+            {
+                const string errorMessage = "Uri query contains invalid params";
+                Debug.LogError(errorMessage);
+                throw new ArgumentException(errorMessage);
+            }
+
+            return GetQueryParamsFromUri(uri).ToArray()[0];
+        }
+
+        /// <summary>
+        /// Inspect the URI and determine whether it contains valid params according to the following policies:
+        ///   1. URI query must include "code" or "error" in param keys.
+        ///   2. "code" and "error" cannot be present at the same time.
+        ///   3. No other keys apart from "code", "error" are allowed.
+        /// </summary>
+        internal static bool UriContainsValidQueryParams(Uri uri)
+        {
+            var allowedQueries = new[] {"code", "error"};
+            var queryParams = GetQueryParamsFromUri(uri);
+
+            Predicate<Dictionary<string, string>> codeOrErrorIsPresent = queryParamsDict =>
+                queryParamsDict.ContainsKey("code") || queryParamsDict.ContainsKey("error");
+
+            Predicate<Dictionary<string, string>> notBothCodeAndErrorArePresent = queryParamsDict =>
+                !(queryParamsDict.ContainsKey("error") && queryParamsDict.ContainsKey("code"));
+
+            Predicate<Dictionary<string, string>> noOtherKeysArePresent = queryParamsDict =>
+                queryParamsDict.Where(kvp => !allowedQueries.Contains(kvp.Key)).ToArray().Length == 0;
+            return codeOrErrorIsPresent(queryParams) && notBothCodeAndErrorArePresent(queryParams) &&
+                   noOtherKeysArePresent(queryParams);
+        }
+
+        /// <summary>
+        /// Processes URI, extracts query params, puts them into a dictionary returns the dictionary.
+        /// </summary>
+        internal static Dictionary<string, string> GetQueryParamsFromUri(Uri uri)
+        {
+            var result = new Dictionary<string, string>();
+            foreach (var pair in uri.Query.Substring(1).Split('&'))
+            {
+                if (pair.Contains("="))
+                {
+                    var keyAndValue = pair.Split('=');
+                    result.Add(keyAndValue[0], keyAndValue[1]);
```

Comment:
Do you need to unescape these strings?
https://docs.unity3d.com/ScriptReference/WWW.UnEscapeURL.html

Potentially not if the Google auth API guarantees that content is already URL safe. If this is the case, add a comment saying so.